### PR TITLE
feat(layouts): decompose and centralize layout views

### PR DESCRIPTION
**Pull request by hypeJunction**
(replaced by #10333)

**feat(layouts): decompose and centralize layout views**

Centralizes layouts in a default layout providing better control over which layout elements should be rendered. Decomposes default layout view into smaller elements.

BREAKING CHANGE
one_column, one_sidebar, two_sidebar and content layout views have been removed - plugins should start using the default layout. Plugins need to update their use of page/layouts/default and
page/layouts/elements/filter.

**feat(groups): update group listings to new layout API** 

BREAKING CHANGE
groups/groups_sort_menu view has been removed. Plugins that were using it to alter filter tabs on groups/all page, should instead use "register, menu:filter:groups/all" plugin hook.

### CODE REVIEW

**Comment by hypeJunction** [comment]
So, here is a radical idea! 💩

File: docs/guides/upgrading.rst [hunk 1]
```diff
@@ -162,6 +170,21 @@ If handlers were registered with the same priority, these are called in the orde
...
+New approach to page layouts
+----------------------------
+
+Page layouts have been decomposed into smaller elements, with each layout element acquiring its wrapper div. This should make it easier for theme to target specific layout elements without having to override layouts at large.
+
+``one_column``, ``one_sidebar``, ``two_sidebar`` and ``content`` layouts have been removed - instead layout rendering has been centralized in the ``default`` layout with full controls over what parts of the layout should be rendered via ``$vars``.
+This takes away control over visual presentation from plugins and hands that control over to the themes. For maximum backwards compatibility, calls to ``elgg_view_layout()`` with these layout names will still yield expected output, but the plugins should start using the ``default`` layout.
+Themes overriding any of the above layout views have to be updated to properly render ALL layout elements.
+
+Themes and plugins overriding ``layouts/elements/filter`` view has to update their usage of view parameters.
```
Comment (hypeJunction) [inline docs/guides/upgrading.rst]: `page/layouts/elements/filter`
have to

File: views/default/page/layouts/error.php [hunk 2]
```diff
@@ -7,11 +7,8 @@
...
-$class = 'elgg-layout-error';
-if (isset($vars['class'])) {
-	$class = "$class {$vars['class']}";
-}
-$vars['class'] = $class;
+$vars['class'] = elgg_extract_class($vars, 'elgg-layout-error');
+
 $vars['nav'] = ''; // prevents outputting breadcrumbs
 
 echo elgg_view('page/layouts/one_column', $vars);
```
Comment (hypeJunction) [inline views/default/page/layouts/error.php]: Update to use `default` layout

Comment (hypeJunction) [inline views/default/page/layouts/error.php]: No sidebar

File: views/default/page/layouts/default.php [hunk 3]
```diff
@@ -1,36 +1,58 @@
...
+	$class[] = 'elgg-layout-one-column';
 }
 
-echo elgg_view('page/layouts/elements/footer', $vars);
+$body = elgg_view('page/layouts/elements/body', $vars);
 
-echo '</div>';
+echo elgg_format_element('div', [
+	'class' => $class,
+], $sidebar . $sidebar_alt . $body);
```
Comment (hypeJunction) [inline views/default/page/layouts/default.php]: In Aalborg, it's `$sidebar_alt . $body . $sidebar`, but if we are switching to flexbox, we can just use `order`

File: docs/guides/upgrading.rst [hunk 4]
```diff
@@ -162,6 +170,21 @@ If handlers were registered with the same priority, these are called in the orde
...
+
+As a result of these changes, all layouts can now be easily extended to have multiple tabs. Plugins can pass ``filter_id`` parameter that will allow other plugins to hook into ``register, menu:filter:<filter_id>`` hook and add new tabs. If no ``filter_id`` is provided, default ``register, menu:filter`` hook can be used.
+
+Layout views and their underlying views now receive ``identifier`` and ``segments`` of the page being rendered, so they can make decisions about what elements should be rendered.
```
Comment (mrclay) [inline docs/guides/upgrading.rst]: So we're asking these views to do their own routing?

File: engine/lib/views.php [hunk 5]
```diff
@@ -693,31 +693,88 @@ function elgg_view_layout($layout_name, $vars = array()) {
 	}
 	$timer->begin([__FUNCTION__]);
 
-	$params = array();
```
Comment (mrclay) [inline engine/lib/views.php]: Updating docs of this function might help get this idea across

File: views/default/page/layouts/elements/sidebar.php [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+ */
+
+$sidebar = elgg_extract('sidebar', $vars);
+if ($sidebar === false) {
+	return;
+}
+?>
+<div class="elgg-sidebar elgg-layout-sidebar clearfix">
+	<?= elgg_view('page/elements/sidebar', $vars) ?>
+</div>
```
Comment (mrclay) [inline views/default/page/layouts/elements/sidebar.php]: Curious, why not just add the wrapper to page/elements/sidebar? A consistency thing?

File: docs/guides/upgrading.rst [hunk 7]
```diff
@@ -162,6 +170,21 @@ If handlers were registered with the same priority, these are called in the orde
...
+
+``one_column``, ``one_sidebar``, ``two_sidebar`` and ``content`` layouts have been removed - instead layout rendering has been centralized in the ``default`` layout with full controls over what parts of the layout should be rendered via ``$vars``.
+This takes away control over visual presentation from plugins and hands that control over to the themes. For maximum backwards compatibility, calls to ``elgg_view_layout()`` with these layout names will still yield expected output, but the plugins should start using the ``default`` layout.
+Themes overriding any of the above layout views have to be updated to properly render ALL layout elements.
+
+Themes and plugins overriding ``layouts/elements/filter`` view has to update their usage of view parameters.
+
+As a result of these changes, all layouts can now be easily extended to have multiple tabs. Plugins can pass ``filter_id`` parameter that will allow other plugins to hook into ``register, menu:filter:<filter_id>`` hook and add new tabs. If no ``filter_id`` is provided, default ``register, menu:filter`` hook can be used.
```
Comment (mrclay) [inline docs/guides/upgrading.rst]: 👍

**Comment by mrclay** [comment]
If I follow, if I supply content to `sidebar_alt` I get the 2nd column automatically. I wonder why we didn't do it that way before.

File: docs/guides/upgrading.rst (hunk 4 above)
Comment (mrclay) [inline docs/guides/upgrading.rst]: ...Allowing them. That's fine.

**Comment by jdalsem** [comment]
Finally... one layout to rule them all...

File: views/default/page/layouts/elements/sidebar.php (hunk 6 above)
Comment (hypeJunction) [inline views/default/page/layouts/elements/sidebar.php]: This seemed cleaner

**Comment by hypeJunction** [comment]
- Updated the docs
- Threw before and after layout hooks out of the window
- Updated admin/error layouts for consistency
- Updated theme sandbox with examples

File: mod/developers/views/default/theme_sandbox/layouts/two_sidebar.php [hunk 8]
```diff
@@ -1,9 +1,24 @@
...
+			'text' => 'Tab 1',
+			'href' => elgg_http_add_url_query_elements($page_url, ['filter' => 'tab1']),
+		],
+		'tab2' => [
+			'text' => 'Tab 2',
+			'href' => elgg_http_add_url_query_elements($page_url, ['filter' => 'tab2']),
+		],
+	],
+	'filter_value' => get_input('filter', 'tab1'),
+));
```
Comment (mrclay) [inline mod/developers/views/default/theme_sandbox/layouts/two_sidebar.php]: Awesome.

**Comment by mrclay** [comment]
> Threw before and after layout hooks out of the window

Fair enough.

File: views/default/page/layouts/default.php [hunk 9]
```diff
@@ -1,36 +1,71 @@
...
+
 /**
  * Elgg default layout
  *
- * @package Elgg
- * @subpackage Core
+ * @uses $vars['class']        Additional CSS classes to apply to the layout
+ *
+ * @uses $vars['breadcrumbs']  An array of breadcrumbs (see navigation/breadcrumbs)
+ * @uses $vars['nav']          Alternative page navigation (replaces navigation/breadcrumbs)
```
Comment (hypeJunction) [inline views/default/page/layouts/default.php]: Should we throw this out? If breadcrumbs is string use it, if array - render navigation/breadcrumbs?

Comment (mrclay) [inline views/default/page/layouts/default.php]: SGTM

**Comment by mrclay** [comment]
Do you feel this is approaching "finished"? Should I convert one plugin to use this to get the feel of it?

**Comment by hypeJunction** [comment]
I feel it's done. There typos in docs that need fixing but I don't see them anymore. Please do adapt something.  River could use some love.

**Comment by hypeJunction** [comment]
I have some more plans for layouts/pages, so the sooner we merge this, the better.

**Comment by hypeJunction** [comment]
@mrclay Could you make this one a priority, so we can merge it sooner?

Once this is in, I want to try the following:

- Update `elgg_view_page()` to accept an array of layout parameters and teach it to draw layouts:

```php
$params = [
     // Shell params
    `#layout` => 'default',
    '#title' => 'Page title',
    '#shell' => 'default',
    // Layout params + shell vars
    'title' => 'Layout title',
    'content' => 'Layout content',
    'entity' => $entity,
    'page_owner' => $page_owner,
];

echo elgg_view_page($params);
// equals to
echo elgg_view_resource('default', $params);
```

File: engine/lib/navigation.php [hunk 10]
```diff
@@ -272,31 +272,50 @@ function elgg_pop_breadcrumb() {
...
 	$params = array(
 		'breadcrumbs' => $breadcrumbs,
 	);
+
+	$params['identifier'] = _elgg_services()->request->getFirstUrlSegment();
+	$params['segments'] = _elgg_services()->request->getUrlSegments();
+	array_shift($params['segments']);
+
 	$breadcrumbs = elgg_trigger_plugin_hook('prepare', 'breadcrumbs', $params, $breadcrumbs);
+	_elgg_services()->logger->error('"prepare, breadcrumbs" hook must return an array of breadcrumbs');
```
Comment (mrclay) [inline engine/lib/navigation.php]: Move this inside the condition

**Review by mrclay — APPROVED** [review]
Besides moving the error line, I'm ready to go for it.

**Comment by mrclay** [comment]
Replaced by #10333

**Pull request closed without merging**